### ISSUE TITLE
py-spy FreeBSD 12 Error

### ISSUE THREAD

**Issue by streaml1ne555** [benfred/py-spy#145]
Getting this error:

Error: Device busy (os error 16)

When running this sequence:

py-spy record --format speedscope --output sample.speedscope -- python2.7 ./sample.py

Fairly generic error and I don't see anything printed out elsewhere. This is on 12.0-RELEASE using the binary release of 0.2.0-dev3 from the releases page.

**Comment by benfred** [comment]
I just tried this out (using vagrant w/ the “freebsd/FreeBSD-12.0-STABLE” box) - and it worked for me =(.

There isn’t much to go on with the error message, but I’m wondering if py-spy is failing to suspend the process to take a stack trace. Can you try again passing in the ```--nonblocking``` option?

Also can you try enabling logging and paste the output here (```RUST_LOG=info py-spy record …```) and maybe even trying to get a backtrace here (same thing but setting ```RUST_BACKTRACE=1``` instead).

Finally, using the binary from the 0.2.0.dev3 release has some issues : https://github.com/benfred/py-spy/issues/141 . I’d recommend compiling from source until I can push out a new release

**Comment by kxepal** [comment]
Can reproduce this with 0.2.0.dev3 and current master on 11.2, but it works only with Python 2.7 - spied process get killed. `--nonblocking` option leads to the same result. Suddenly, both `RUST_LOG` (info or debug) and `RUST_BACKTRACE` shows no additional information, except `Error: Device busy (os error 16)`. 

However, for some runs, py-spy from master was able to record process samples successfully, but in the end (Control+C) tracked process was also killed.

**Comment by benfred** [comment]
I took another look at this problem again and managed to reproduce it with python 2.7 on freebsd (python 2.7 installed from pkg )

Without ```--nonblocking``` profiling all the python2.7 programs I tried failed, with the logs looking like
```
[2019-08-07T10:43:42Z INFO  py_spy::python_spy] Failed to connect to process, retrying. Error: Failed to suspend process
[2019-08-07T10:43:42Z INFO  py_spy::python_spy] Failed to connect to process, retrying. Error: Device busy (os error 16)
[2019-08-07T10:43:42Z INFO  py_spy::python_spy] Failed to connect to process, retrying. Error: Device busy (os error 16)
[2019-08-07T10:43:42Z INFO  py_spy::python_spy] Failed to connect to process, retrying. Error: Device busy (os error 16)
[2019-08-07T10:43:42Z INFO  py_spy::python_spy] Failed to connect to process, retrying. Error: Device busy (os error 16)
[2019-08-07T10:43:42Z INFO  py_spy::python_spy] Failed to connect to process, retrying. Error: Device busy (os error 16)

Error: Device busy (os error 16)
```
with the ```--nonblocking``` option set everything seems to works fine. Profiling the same programs using 3.6  seem to work with or without the --nonblocking option (which is what I was originally testing).

@akhramov Do you have any ideas here?  For the python2.7 case, it looks like it failed to suspend the process once and didn't recover.

There also seems to be another problem that py-spy doesn't recognize if the profiled process has finished running or not on freebsd - profiling short lived processes just start causing errors in py-spy after the profiled process exits.

@kxepal When you say that ```but in the end (Control+C) tracked process was also killed.``` - are you getting py-spy to create the process or connecting to another process using the --pid flag?  Under normal operations we kill any subprocesses that py-spy creates after profiling, but leave processes that py-spy doesn't create alone:

https://github.com/benfred/py-spy/blob/755ba04e11e18bb039e0c16e594b09d29ab0b699/src/main.rs#L402-L403

**Comment by akhramov** [comment]
> For the python2.7 case, it looks like it failed to suspend the process once and didn't recover.

Sounds like the code is reporting that we were not able to lock the process via ptrace, whereas in reality we did lock the process: os error 16, EBUSY, means just that: someone is tracing the process.

If `py-spy` does not release the lock and exits, the profiled process is going to be killed, per ptrace(2) specification.

> There also seems to be another problem that py-spy doesn't recognize if the profiled process has finished running or not on freebsd - profiling short lived processes just start causing errors in py-spy after the profiled process exits.

Is this observed on python2.7 only? I have a feeling that this might be caused by the locks problem as well.

I will take a closer look at our locking mechanics and will get back to you. Please don't hesitate involving me in FreeBSD-related bugs as early as possible: I'm really happy to help you.

**Comment by benfred** [comment]
> Sounds like the code is reporting that we were not able to lock the process via ptrace, whereas in reality we did lock the process: os error 16, EBUSY, means just that: someone is tracing the process.

Yeah - thats sort of my take here as well. I'm wondering if we need to call waitpid multiple times if WIFSTOPPED doesn't return true. Fwiw, pyflame calls waitpid in a loop until WIFSTOPPED here https://github.com/uber/pyflame/blob/master/src/ptrace.cc#L40-L47 (thats on linux, but believe all these calls are a posix standard so the same logic might apply).

> Is this observed on python2.7 only? I have a feeling that this might be caused by the locks problem as well.

I saw this on both python 2.7 and 3.6. Running this command:

```
py-spy top -- python -c 'import time; time.sleep(2)'
```

shows errors after 2 seconds ('failed to suspend process' when run without the nonblocking option, and 'Failed to copy PyInterpreterState from process' when run with nonblocking option). Running  this same command on OSX/linux and py-spy exits when the profiled process finishes

**Comment by akhramov** [comment]
So, to summarize my findings.


1. The locks problem (observed in 2.7)

Actually, we have a bug in `read-process-memory` crate. Upon failed `PTRACE_IO` call, we do not detach from the process therefore `remoteprocess` crate unable to obtain the lock:

https://github.com/benfred/read-process-memory/blob/8bbcb7a52e24588e005f62c5837cd323bbb6c930/src/lib.rs#L367-L370

Why does `PTRACE_IO` fail? 

I narrowed the problem down to this line.

https://github.com/benfred/py-spy/blob/a481138c2849461e593dcd6095641879d256bcf5/src/python_spy.rs#L222

`threads` contains

```
{ next: 0x27, interp: 0x80041b788, frame: 0x0, recursion_depth: 2657530, tracing: 8, use_tracing: 872, c_profilefunc: Some(0x28), c_tracefunc: Some(0x800366e30), c_profileobj: 0x0, c_traceobj: 0x0, curexc_type: 0x0, curexc_valu
e: 0x0, curexc_traceback: 0x800366f90, exc_type: 0x0, exc_value: 0x0, exc_traceback: 0x0, dict: 0x80034e3d0, tick_counter: 3567872, gilstate_counter: 8, async_exc: 0x800368a80, thread_id: 34363306576, trash_delete_nesting: 3568624,
 trash_delete_later: 0x0 }   
```

On the next iteration, we are trying  to read from `0x27`. Obviously, we fail. The struct layout seems to be correct. @benfred, does this have something to do with the interpreter address?

2. Short running processes...

`exe()` returns the empty string for non-running processes:

https://github.com/benfred/py-spy/blob/a25648b889f2a313284237b7014265c1c8115a53/remoteprocess/src/freebsd/mod.rs#L52-L54

The underlying `libprocstat` library must have a way to identify whether proc is stopped or not. Otherwise we can employ a workaround:

```rust
    pub fn exe(&self) -> Result<String, Error> {
        let filename = procstat::exe(self.pid)?;

        if filename.is_empty() {
            return Err(
                Error::Other("Failed to get process executable name".into())
            );
        }

        Ok(filename)
    }
```

3. On ptrace & waitpid

I honestly think we do not have to have a loop. lldb doesn't
https://github.com/freebsd/freebsd/blob/56c04b0bcfcd116f1b13087ec13bcba2d8dc7705/contrib/llvm/tools/lldb/source/Plugins/Process/FreeBSD/ProcessMonitor.cpp#L1008-L1020

**Comment by benfred** [comment]
@akhramov nice work narrowing this down!

> On the next iteration, we are trying to read from 0x27. Obviously, we fail. The struct layout seems to be correct.  @benfred, does this have something to do with the interpreter address?

On freebsd, it doesn't seem like we can get the location of the interpreter from symbols - so we're relying on scanning the BSS section to find a pointer to the python intrepreter:

```
[2019-08-08T09:32:17Z INFO  py_spy::python_spy] Got virtual memory maps from pid 8109:
[2019-08-08T09:32:17Z INFO  py_spy::python_spy] Found libpython binary @ /usr/local/lib/libpython3.6m.so.1.0
[2019-08-08T09:32:17Z INFO  py_spy::python_spy] Getting version from python binary BSS
[2019-08-08T09:32:17Z INFO  py_spy::python_spy] Failed to get version from BSS section: failed to find version string
[2019-08-08T09:32:17Z INFO  py_spy::python_spy] Getting version from libpython BSS
[2019-08-08T09:32:17Z INFO  py_spy::version] Found matching version string '3.6.9 (default, Jul 25 2019, 01:18:58) '
[2019-08-08T09:32:17Z INFO  py_spy::python_spy] python version 3.6.9 detected
[2019-08-08T09:32:17Z INFO  py_spy::python_spy] Failed to get interp_head from symbols, scanning BSS section from main binary
[2019-08-08T09:32:17Z INFO  py_spy::python_spy] Failed to get interpreter from binary BSS, scanning libpython BSS
[2019-08-08T09:32:17Z INFO  py_spy::python_spy] Found interpreter at 0x00000008009eb000
```
The way we scan the BSS section for the interpreter state - is that we're looking for valid pointers to a PyInterpreterState that have a memory layout like we expect (specifically the  pyinterpretstate object has a tstate_head memory that points to a PyThreadState object, which has an interp member that should point back to the same PyInterpreterState object - so we just check that these members are all valid pointers and are set up like we expect).

The final check for figuring out if we have a valid interpreter address is checking to see if we can get a stacktrace from the interpreter at that address: https://github.com/benfred/py-spy/blob/755ba04e11e18bb039e0c16e594b09d29ab0b699/src/python_spy.rs#L524-L534

It seems like for the python2.7 interpreter I was testing on this, this check gets called with something that isn't a valid PyInterpreterState , and then we don't detach from the process appropiately causing the issues above (and its killed when py-spy exits).

To test out if this is the problem, I built a version of py-spy with this  fix https://github.com/benfred/read-process-memory/pull/3 . This seems to fix the issue with python 2.7 for me anyways.

I think we should make a couple of other change here too though:
 * the return call from the ptrace_attach call probably should be an ```Option<PtraceLock>```, which has a drop trait to automatically call ptrace_detach when it goes out of scope
 * the py-spy nonblocking option makes no sense on freebsd, since we attach/block with each read process memory call. We should disable it on freebsd
 * I'm wondering if we should refactor remoteprocess so that we don't have to call ptrace attach on every read_process_memory call, just to have ptrace return that we are already attached (would be a decent amount of work - and is probably out of scope of the immediate fixes here)

> The underlying libprocstat library must have a way to identify whether proc is stopped or not. Otherwise we can employ a workaround:

The workaround you proposed seems fine to me (though we probably want to have a 'running' check eventually in remoteprocess, rather than relying on the ```exe``` method returning an error.  I've created a new issue to track this here https://github.com/benfred/py-spy/issues/146

>    On ptrace & waitpid I honestly think we do not have to have a loop. lldb doesn't

Thanks for the lldb link! (py-spy also don't have a loop with ptrace in linux fwiw - I've just never been sure that we weren't missing something here)

**Comment by benfred** [comment]
Added the read_process_memory fix to master, and is also in the 0.2.0.dev4 release - you will be able to download a freebsd build from the releases page soon.

I've also applied some other FreeBSD related fixes here https://github.com/benfred/py-spy/pull/149 . This includes a change to locking the process while initializing the pythonspy object (that would have also fixed this problem, albeit quite a bit more indirectly) and fixes to detect when the process exits on FreeBSD.
